### ISSUE TITLE
"sp-pnp-js" package: "^ 2.0.0" or 2.0.1" SCRIPT1002: Syntax Error in explorer 10 

### ISSUE THREAD

**Issue by danielseis** [SharePoint/PnP-JS-Core#327]
Good afternoon to all the people here active.
I need somebody to help me.
I'm starting **with angular2**, my problem has come by incorporating sp-pnp-js into the project. Everything progressed properly until the customer would have it in explorer 10..11 ... surprise!
it does not work.
I've been trying to do / see compatibility for 3 days and I can not get anything.
The error of the explorer: SCRIPT1002: Syntax error in the file **vendor.bundle.js**
**Polyfills.bundle.js** 

As soon as I compile the project with the "sp-pnp-js" package: "^ 2.0.0", explorer stops working.

Any suggestions

thank you very much.

**Comment by patrick-rodgers** [comment]
Are there any specific errors reported in the console? Without more detailed information I am just guessing, but I think you likely need the polyfills for promises and fetch as [described here](https://github.com/SharePoint/PnP-JS-Core/wiki/Getting-Started:-Install-&-Use#polyfill). IE 10 & 11 don't support those natively - so if your application is working in edge or chome then this would be the first thing I'd check.

**Comment by danielseis** [comment]
#### Category

[X] Error

#### Expected Behavior

When runs an Angular 2 app in IE10 or IE11 with the required polyfills for pnp (**core-js/es6/promise** and **whatwg-fetch**) the application crash.

#### Observed Behavior

The next error appears:
```
SCRIPT1002: Syntax Error
Vendor.bundle.js
Const decorators_1 = __webpack_require __ (876);

SCRIPT5007: Can not get null or undefined 'call' reference property
```


When I install `sp-pnp-js version 1.0.6` the application works as spected in both IE10 and IE11.
Using `sp-pnp-js 2.0.0` or `2.0.1` the same application **does not work in IE10 nor IE11**.


#### Steps to Reproduce

* Setup a sample Angular 2 project:
```
> npm install -g angular-cli
> ng new test-pnp
> cd test-pnp
> npm install sp-pnp-js --save
```

* Navigate to the project folder and open with your favorite editor (e.g. Visual Studio Code):

```
> code .
```

* Edit the Angular 2 main component **app.component.ts**, import the pnp library and make a call to some pnp method (e.g. setup):

```
import { Component } from '@angular/core';

import * as pnp from "sp-pnp-js";


@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent {
  title = 'app works!';

    constructor() {

        // Sets up the pnp library.
        pnp.setup({
            headers: {
                "Accept": "application/json; odata=verbose"
            }
        });

    }
}
```

* Serve the application using the Angular-Cli commands:
```
> ng serve
```

* Open the browser IE10 or IE11 and show console to see the error mentioned.


Some help will be very appreciated as we have the project bloked in a customer due to this issue.
Thanks in advance.

**Comment by patrick-rodgers** [comment]
If angular 2 doesn't handle processing es6 bundles (guessing that is the issue since you said 1.0.6 worked) you could look at either adding [babel](https://babeljs.io/) into your build chain - or marking the library as external so it isn't built into the project. We discuss that in relation to SPFx [here](https://github.com/SharePoint/PnP-JS-Core/wiki/Deployment#cdn-and-spfx).

**Comment by danielseis** [comment]
I'm still with the problem.

```
> npm install -g @angular/cli
> ng new PROJECT_pnp
> cd PROJECT_pnp
> npm install sp-pnp-js --save
> npm install es6-promise --save
> npm install whatwg-fetch --save

```
my polyfills:

```
/** IE9, IE10 and IE11 requires all of the following polyfills. **/
import 'core-js/es6/symbol';
import 'core-js/es6/object';
import 'core-js/es6/function';
import 'core-js/es6/parse-int';
import 'core-js/es6/parse-float';
import 'core-js/es6/number';
import 'core-js/es6/math';
import 'core-js/es6/string';
import 'core-js/es6/date';
import 'core-js/es6/array';
import 'core-js/es6/regexp';
import 'core-js/es6/map';
import 'core-js/es6/set';

/** IE10 and IE11 requires the following for NgClass support on SVG elements */
import 'classlist.js';  // Run `npm install --save classlist.js`.

/** IE10 and IE11 requires the following to support `@angular/animation`. */
import 'web-animations-js';  // Run `npm install --save web-animations-js`.


/** Evergreen browsers require these. **/
import 'core-js/es6/reflect';
import 'core-js/es7/reflect';


/** ALL Firefox browsers require the following to support `@angular/animation`. **/
import 'web-animations-js';  // Run `npm install --save web-animations-js`.

/********************************
 * section https://github.com/SharePoint/PnP-JS-Core/wiki/Install-and-Use
 */
import 'es6-promise/dist/es6-promise';
import 'whatwg-fetch';

/***************************************************************************************************
 * Zone JS is required by Angular itself.
 */
import 'zone.js/dist/zone';  // Included with Angular CLI.
```

my package.json

```
"dependencies": {
    "@angular/common": "^2.4.0",
    "@angular/compiler": "^2.4.0",
    "@angular/core": "^2.4.0",
    "@angular/forms": "^2.4.0",
    "@angular/http": "^2.4.0",
    "@angular/platform-browser": "^2.4.0",
    "@angular/platform-browser-dynamic": "^2.4.0",
    "@angular/router": "^3.4.0",
    "classlist.js": "^1.1.20150312",
    "core-js": "^2.4.1",
    "es6-promise": "^4.0.5",
    "rxjs": "^5.1.0",
    "sp-pnp-js": "^2.0.1",
    "web-animations-js": "^2.2.2",
    "whatwg-fetch": "^2.0.2",
    "zone.js":  #"^0.7.6"
  },
  "devDependencies": {
    "@angular/cli": "1.0.0-beta.32.3",
    "@angular/compiler-cli": "^2.4.0",
    "@types/jasmine": "2.5.38",
    "@types/node": "~6.0.60",
    "codelyzer": "~2.0.0-beta.4",
    "jasmine-core": "~2.5.2",
    "jasmine-spec-reporter": "~3.2.0",
    "karma": "~1.4.1",
    "karma-chrome-launcher": "~2.0.0",
    "karma-cli": "~1.0.1",
    "karma-jasmine": "~1.1.0",
    "karma-jasmine-html-reporter": "^0.2.2",
    "karma-coverage-istanbul-reporter": "^0.2.0",
    "protractor": "~5.1.0",
    "ts-node": "~2.0.0",
    "tslint": "~4.4.2",
    "typescript": "~2.0.0"
  }
```
my component app.component.ts

```
import { Component } from '@angular/core';

import * as pnp from "sp-pnp-js";

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})

export class AppComponent {
  title = 'app works!';

  constructor() {

    // Sets up the pnp library.
    pnp.setup({
      headers: {
        "Accept": "application/json; odata=verbose"
      }
    });
  }
}

```

file tsconfig.json

```
{
  "compilerOptions": {
    "baseUrl": "",
    "declaration": false,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "lib": [
      "es2016",
      "dom"
    ],
    "mapRoot": "./",
    "module": "es2015",
    "moduleResolution": "node",
    "outDir": "../dist/out-tsc",
    "sourceMap": true,
    "target": "es5",
    "typeRoots": [
      "../node_modules/@types"
    ]
  }
}
```

then 

`> ng serve
`

Ie10 and ie11 not work.

SCRIPT1002: Error de sintaxis

**Comment by patrick-rodgers** [comment]
I still think this is an issue with angular pulling in the es6 modules directly. I am not super familiar with angular 2 so will need to find time to setup a project and test this. Does it work in other browsers?

**Comment by johnnliu** [comment]
webpack's modules looks pretty strange.

![image](https://cloud.githubusercontent.com/assets/1279372/23192356/18c16a60-f8f7-11e6-8e1a-2c7dffe3d1a2.png)

**Comment by patrick-rodgers** [comment]
I took some time tonight to recreate this and can confirm that it is an issue related to the es6 modules being pulled directly and added to the bundle by the default angular configuration. I don't know angular well enough off the top of my head to suggest a fix, but I'll do a little research. @danielseis - have you tried anything that might give us a clue?

**Comment by johnnliu** [comment]
spent the evening catching up and playing with this.  some notes.

when sp-pnp-js is included, `ng build` pulls through individual files from node_modules.  

![image](https://cloud.githubusercontent.com/assets/1279372/23211830/046e0e90-f958-11e6-9d1b-3efb316b8be6.png)

`sp-pnp-js/lib/utils/util.js` seems to have been built to ES6, with keywords such as 'class' in the plain script file.  

Since it is JS file - the typescript compiler does not try to turn it into ES5, and webpack happily packs it into the vendor.js

![image](https://cloud.githubusercontent.com/assets/1279372/23211889/4d0f6c0c-f958-11e6-9358-36de108829fd.png)

This works fine in Chrome/Edge, but dies horribly on IE11, which complains that 'class' is bad syntax.

![image](https://cloud.githubusercontent.com/assets/1279372/23211927/675de372-f958-11e6-9ab7-d84393604ed7.png)

After which, IE11 stops parsing the rest of the file, leaving several holes in the webpack modules array.

We see a second error later when webpack tries to call the modules and several of them are empty, leading to the null.call() error.

**Comment by johnnliu** [comment]
in the latest `@angular/cli` we can do `ng eject` which will split the webpack.config from the ng build process.  I'm playing with this to add an uglify plugin, which can re-fix the bundled JS code before it gets shuffled into vendor.js

**Comment by patrick-rodgers** [comment]
@johnnliu - awesome work! @danielseis - if you follow the steps above does this work for you? If so we can roll it up into a wiki article in case others have the same issue.

**Comment by johnnliu** [comment]
Hey @patrick-rodgers is there a strong reason that we need to target ES2015 (ES6) in tsconfig rather than say ES5?  If we target ES5 we can have ES5 javascript in node_modules.

I think this is more philosophical regarding whether we want to support only Evergreen Browsers.  But alienating IE11 seems a bit harsh particularly given the enterprise SharePoint audience.  Polyfills will work for missing properties but it dies horribly on 'class' with syntax error.  Can't polyfill that...

**Comment by patrick-rodgers** [comment]
@johnnliu - the distribution is es5 (files in dist). There are advantages to having es6 modules in lib (bundle optimization [rollup, webpack2]) and edge, chrome, firefox support it if you build it in directly. This was a big decision with the 2.0.0 release and I think correct. We also have the es5 compat dist files on cdnjs so folks can use it as an external very easily. Angular 2 is a bit of a special case as it "hides" everything so you can't update the setting without eject.

TL:DR: Not trying to alienate IE 11 (dist files are es5) but there is value in having the lib be es6.

**Comment by johnnliu** [comment]
ok, no problems.  I'm quite happy either way.

actually, with `ng eject` and taking back control of webpack.config, we can use webpack externals as well.  That may be suitable as a second alternative solution that uses the ES5 compat dist file.   Would be useful if using Angular with SPFx.

**Comment by danielseis** [comment]
@patrick-rodgers - Yes, I followed the steps and it works perfectly! I'm really grateful, @johnnliu. 

Lucky that angular cli has been updated to be able to perform the `ng eject` !.

Making tests, we can dispense with the babel-loader

Web.pack.config

```
      {
        "test": /\.ts$|sp-pnp-js\/lib.*\.js$/,
        "loader": "@ngtools/webpack",
        "exclude": [
          /\.(spec|e2e)\.ts$/
        ]
      },
      // {
      //   "test": /sp-pnp-js.*\.js$/,
      //   "loader": "babel-loader?presets[]=es2015"
      // }
```

At present and with the versions of packages that we are working on, what has been key is that @johnnliu has detected what was compiled and what was not compiling.

> Since it is JS file - the typescript compiler does not try to turn it into ES5, and webpack happily packs it into the vendor.js

I give the incident closed, many thanks!

**Comment by patrick-rodgers** [comment]
Great news @danielseis ! Glad you are unblocked!

And thanks @johnnliu - great work on finding the fix.

I'll leave this open as a note to myself to create a wiki entry on this topic.

**Comment by johnnliu** [comment]
I really like how @danielseis just use `@ngtools/webpack` to re-pack sp-pnp-js.  I had further experimented with using ts-loader instead of babel-loader last night but `@ngtools` was quite unhappy with a second typescript loader in the mix.  It didn't occur to me I can just tell `@ngtools` to take care of it.

So that solution kicks ass!

**Comment by mpowney** [comment]
Just wanted to note a similar issue with a project I worked on last week. It's a NodeJS project with a dependency on "sp-pnp-js": "^2.0.1", "react": "^15.4.2", bundling with webpack, and transpiling with babel.

IE in v10 mode would encounter a Syntax Error when it came across a line with "const" (an ES2015 declaration) from the sp-pnp-js module.

Reverting to "sp-pnp-js": "^1.0.4", the IE10 Syntax Error would go away, with all other package.json and webpack.config.js config the same.

Changing my webpack.config.js to find the sp-pnp-js/*.js files and running them explicitly through babel's es2015 loader (similar to the solution above) does the trick.  My webpack.config.js file now looks like this:

```javascript
var path = require('path');
var webpack = require('webpack');
var CopyWebpackPlugin = require('copy-webpack-plugin');

module.exports = {
    context: path.join(__dirname, '/src'),
    entry: {
        "sppnpes2015": ['./scripts/main.js'],
        vendor: ['babel-polyfill', 'react', 'react-dom', 'sp-pnp-js']
    },
    output: {
        path: path.join(__dirname, '/dist/style.library/@sppnpes2015'),
        filename: '/scripts/[name].js',
        library: 'sppnpes2015',
        libraryTarget: 'var'
    },
    module: {
        loaders: [{
                test: /sp-pnp-js.*\.js?$/,
                loader: 'babel-loader',
                //exclude: /node_modules/,
                query: {
                    presets: ['es2015']
                }
            }, {
                test: /\.js$/,
                exclude: /node_modules/,
                loader: 'babel-loader',
                query: {
                    presets: ['es2015']
                }
            }, {
                test: /\.jsx?$/,
                loader: 'babel-loader',
                exclude: /node_modules/,
                query: {
                    presets: ['es2015', 'react']
                }
            }

        ]
    },
    plugins: [
        new webpack.optimize.CommonsChunkPlugin({
            name: 'vendor',
            filename: 'scripts/vendor.bundle.js'
        }),
        new CopyWebpackPlugin([{
            from: './html',
            to: 'html'
        }])
    ]

};
```

Wondering if this SHOULD happen?  Other libraries in node_modules don't seem to contain ES2015 declarations, Is there an issue here that the nodejs published modules aren't transpiled down (I'm just guessing here as I'm quite new to the world of NodeJS and what conventions are / should be followed)?

**Comment by patrick-rodgers** [comment]
After listening to the feedback from folks there is enough impact to roll back the lib folder to es5. I have merged that to dev and it will be part of 2.0.2 coming on March 10th. Hope that helps people out and removes a pain point.

I do still encourage folks to use the library as an external when packing it up for deployment - but forcing it was a mistake. No need to make anyone do an extra thing to use the library.
